src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/SafeBuffer.cs

```diff
@@ -217,24 +217,31 @@ public void ReadArray<T>(ulong byteOffset, T[] array, int index, int count)
             if (array.Length - index < count)
                 throw new ArgumentException(SR.Argument_InvalidOffLen);
 
+            ReadSpan(byteOffset, new Span<T>(array, index, count));
+        }
+
+        [CLSCompliant(false)]
+        public void ReadSpan<T>(ulong byteOffset, Span<T> buffer)
+            where T : struct
+        {
             if (_numBytes == Uninitialized)
                 throw NotInitialized();
 
             uint sizeofT = SizeOf<T>();
             uint alignedSizeofT = AlignedSizeOf<T>();
             byte* ptr = (byte*)handle + byteOffset;
-            SpaceCheck(ptr, checked((nuint)(alignedSizeofT * count)));
+            SpaceCheck(ptr, checked((nuint)(alignedSizeofT * buffer.Length)));
 
             bool mustCallRelease = false;
             try
             {
                 DangerousAddRef(ref mustCallRelease);
 
-                if (count > 0)
+                if (!buffer.IsEmpty)
                 {
-                    fixed (byte* pStructure = &Unsafe.As<T, byte>(ref array[index]))
+                    fixed (byte* pStructure = MemoryMarshal.AsBytes(buffer))
                     {
-                        for (int i = 0; i < count; i++)
+                        for (int i = 0; i < buffer.Length; i++)
                             Buffer.Memmove(pStructure + sizeofT * i, ptr + alignedSizeofT * i, sizeofT);
```

Comment:
How about using `buffer[i] = Unsafe.As<T>(ptr + alignedSizeofT * i)` here? I'm not sure if it's valid or performant.